README.md

# 依赖python环境

+ python3.5

# 安装说明

执行即可 `pip install MeterOCR-0.0.9-cp35-cp35m-win_amd64.whl`

# 依赖的pip包

+ cv2 (pip install -i https://pypi.tuna.tsinghua.edu.cn/simple opencv-python)
+ matplotlib (pip install -i https://pypi.tuna.tsinghua.edu.cn/simple matplotlib)
+ tensorflow==1.6 (pip install -i https://pypi.tuna.tsinghua.edu.cn/simple tensorflow==1.6)



# api

## 离线调用接口：

```python
from meter_ocr.ifs import Interface

class MeterElecOCR(Interface):
    """A user-created :class:`MeterElecOCR <Interface>` object.

    继承识别类

    Usage::
        >>> from meter_ocr.ifs import Interface
        >>> elec_model = MeterElecOCR()
    """

    def __init__(self):
        super(MeterElecOCR, self).__init__()
        print(self.message)

    def predict(self, img):
        """主要调用接口
        识别待检测图片，并返回字典结果
        img: cv2.Mat 格式图片
        has_det: 是否开启检测模式(适用于大图)，小图请设置为False

        Usage::
            >>> from meter_ocr.ifs import Interface
            >>> elec_model = MeterElecOCR()
            >>> result = elec_model.predict(cv2.UMat)
        """
        return self._predict(img, has_det=True)
```

## 在线调用接口

### POST请求 

`localhost:6002/predict?image=b'abc';has_det='false'`

|名称|类型|必填|说明|
| --- | --- | --- | --- |
|image|base64|True|以base64编码的jpg格式图片|
|has_det|string|False|参数为'True'或'true'时，适用于大图大图识别，否则适用于裁剪后的小图识别|

### 返回参数说明

|名称|类型|说明|
| --- | --- | --- |
|code|int|错误码，非0则返回异常|
|ftime|时间|请求时间戳|
|message|string|错误信息|
|result|字典|包含识别结果|
|time|float|识别耗时|
|bndbox|1d-数组|存储检测区域坐标，格式为：[左上x,左上y,右下x,右下y]|
|polygon|2d-数组|存储斜框区域坐标，格式为：[[左上], [右上], [右下], [左下]]|
|text|string|识别的文本结果|

### JSON 返回示例
```json
{
    "code": 0,
    "ftime": "2019-11-18 21:47:17 PM 1",
    "message": "success",
    "result": {
        "time": 0.031061649322509766,
        "bndbox": [
            59,
            14,
            242,
            50,
            0.9962421655654907
        ],
        "polygon": [
            [
                85.4101791381836,
                56.29902648925781
            ],
            [
                258.4898681640625,
                71.1941146850586
            ],
            [
                258.367431640625,
                91.19723510742188
            ],
            [
                84.28414154052734,
                80.03681945800781
            ]
        ],
        "text": "01673227"
    }
}
```

meter_ocr/CHANGELOG
The list of most significant changes made over time in MeterOCR.

------------------------------------------------------------------------

MeterOCR 0.0.9 (20191104)

Tensorflow Minimum Required (Version 1.6)

Changes:

Preview Features:

- Init repository.

- Det Models Path: /code/disk1/xxx/data/dianbiao/data_v1/ssd/v2/model.pb

- Tps Models Path: /code/disk1/xxx/data/dianbiao/data_v1/four_pts/v3/exports/model_40000.pb

- Rnn Models Path: /code/disk1/xxx/data/dianbiao/data_v1/crnn/v1/exports/model_178000.pb

Bugs fixed:

- None

------------------------------------------------------------------------

MeterOCR 0.0.10 (20191114)

Tensorflow Minimum Required (Version 1.6)

Changes:

- Add Web Services

Preview Features:

- Init repository.

- Det Models Path: /code/disk1/xxx/data/dianbiao/data_v1/ssd/v2/model.pb

- Tps Models Path: /code/disk1/xxx/data/dianbiao/data_v1/four_pts/v3/exports/model_40000.pb

- Rnn Models Path: /code/disk1/xxx/data/dianbiao/data_v1/crnn/v1/exports/model_178000.pb

Bugs fixed:

- resource file path error




meter_ocr/web.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
# 
# Copyright 2016 The TensorFlow Authors All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# @Time    : 2019/11/14 0014 19:44
# @Author  : peichao.xu
# @Email   : xj563853580@outlook.com
# @File    : web.py

# ==============================================================================

from __future__ import absolute_import, division, print_function

import cv2
import time
import json
import base64

import meter_ocr
import socket

import tornado.web
import tornado.gen

import numpy as np

from tornado.concurrent import run_on_executor
from concurrent.futures import ThreadPoolExecutor

from meter_ocr.ifs import Interface

IFS = Interface()

print('OCRService {} at {} (Press CTRL+C to quit) '.format(
    meter_ocr.__version__,
    str(socket.gethostname())))


class MyEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, np.integer):
            return int(obj)
        elif isinstance(obj, np.floating):
            return float(obj)
        elif isinstance(obj, np.ndarray):
            return obj.tolist()
        else:
            return super(MyEncoder, self).default(obj)


class OCRService(tornado.web.RequestHandler):
    executor = ThreadPoolExecutor(50)

    def set_default_headers(self):
        self.set_header('Access-Control-Allow-Origin', '*')
        self.set_header('Access-Control-Allow-Headers', 'x-requested-with')
        self.set_header('Access-Control-Allow-Methods', 'POST, GET, OPTIONS')

    @run_on_executor
    def predict(self):
        info_dict = self.request.body_arguments
        has_det = self.get_argument('has_det', 'false')
        has_det = True if has_det.lower() == 'true' else False

        res = {
            'code': 0,
            'ftime': time.strftime("%Y-%m-%d %H:%M:%S %p %w", time.localtime()),
            'message': 'success',
            'result': {
                'time': 0,
                'bndbox': [],
                'polygon': [],
                'text': []
            },
        }
        try:
            im_base64 = info_dict['image'][0].split(b',')[-1]
            im_bytes = base64.b64decode(im_base64)
            np_array = np.frombuffer(im_bytes, dtype=np.uint8)
            image = cv2.imdecode(np_array, cv2.IMREAD_COLOR)
            pred = IFS._predict(image, has_det=has_det)
            res['code'] = pred['code']
            res['message'] = pred['message']
            res['result']['time'] = pred['time']
            res['result']['bndbox'] = pred['bndbox']
            res['result']['polygon'] = pred['polygon']
            res['result']['text'] = pred['text']
        except Exception as e:
            res['code'] = -1
            res['message'] = str(e)
        res = json.dumps(res, cls=MyEncoder)
        return res

    # @tornado.web.asynchronous
    @tornado.gen.coroutine
    def post(self):
        res = yield self.predict()
        self.write(res)
        self.finish()

    def get(self):
        return self.post()
